google/cloud/internal/log_wrapper.cc

```diff
@@ -36,6 +39,66 @@ std::string DebugString(google::protobuf::Message const& m,
                       (options.single_line_mode() ? " " : "\n"), str, "}");
 }
 
+template <typename T>
+std::string DebugString(google::protobuf::Any const& any,
+                        TracingOptions const& options) {
+  T details;
+  if (!any.UnpackTo(&details)) return {};
+  return DebugString(details, options);
+}
+
+std::string DebugString(Status const& status, TracingOptions const& options) {
+  std::ostringstream os;
+  os << status;
+  auto payload =
+      internal::GetPayload(status, internal::kStatusPayloadGrpcProto);
+  google::rpc::Status proto;
+  if (payload && proto.ParseFromString(*payload)) {
+    // See https://cloud.google.com/apis/design/errors#error_payloads
+    for (google::protobuf::Any const& any : proto.details()) {
+      std::string details;
+      switch (status.code()) {
+        case StatusCode::kInvalidArgument:
+          details = DebugString<google::rpc::BadRequest>(any, options);
```

Comment:
I wonder if we should do this based on the actual contents:

```cc
if (any.Is<google::rpc::BadRequest>()) {
  details = DebugString<google::rpc::BadRequest>(any, options);
} else if (any.Is<google::rpc::PreconditionFailure>()) {
  details = DebugString<google::rpc::PreconditionFailure>(any, options);
}
...
```

Reply:
Unfortunately there appear to be multiple details protos in practice, although in the cases I've seen that the extra protos have been redundant.

For example, a "Not found" error has a `ResourceInfo` message (good), but also a `DebugInfo` message whose `.detail` field just contains the `.DebugString()` of the `ResourceInfo`.

That means in the proposal we would either need to be overly careful with the order of the `Is<M>()` tests, or change `else if` to `if` in that chain and allow for multiple messages in the log, neither of which I preferred.

So, I chose to follow the specification (at least for now), and only use the single details proto that matches the status code.  We could, of course, revisit that if we learn there might be additional, useful information in any extra messages.